crates/core/src/view/home/mod.rs

```diff
@@ -0,0 +1,303 @@
+            top_pos = home.children[home.children.len() - 1].rect().max.y;
+            let row_height = home.children[home.children.len() - 1].rect().height() as i32;
+            fav_index = fav_index + 1;
+
+            // If the next fave would overlap wth the bottom bar, we should not create
+            // any more faves
+            if top_pos + row_height > bottom_bar_top { break };
+        }
+
+        home.set_shelf_index(home.children.len() - 1);
+        home.create_bottom_bar();
+        rq.add(RenderData::new(home.id, rect, UpdateMode::Full));
+        home
+    }
+
+    fn create_background(&mut self) {
+        let bg = Filler::new(self.rect, WHITE);
+        self.children.push(Box::new(bg) as Box<dyn View>);
+    }
+
+    fn create_top_bar(&mut self, format: String, fonts: &mut Fonts, battery: &mut Box<dyn Battery>, frontlight: bool) {
+        let top_bar = TopBar::new(self.rect,
+                                  Event::Toggle(ViewId::SearchBar),
+                                  "Favorite Tags".to_string(),
+                                  format, fonts, battery, frontlight);
+        self.children.push(Box::new(top_bar) as Box<dyn View>);
+    }
+
+    fn create_bottom_bar(&mut self) {
+        let dpi = CURRENT_DEVICE.dpi;
+        let thickness = scale_by_dpi(THICKNESS_MEDIUM, dpi) as i32;
+        let small_height= scale_by_dpi(SMALL_BAR_HEIGHT, dpi) as i32;
+        let (small_thickness, big_thickness) = halves(thickness);
+
+        let separator = Filler::new(rect![self.rect.min.x, self.rect.max.y - small_height - small_thickness,
+            self.rect.max.x, self.rect.max.y - small_height + big_thickness], BLACK);
+        self.children.push(Box::new(separator) as Box<dyn View>);
+        // TODO: should eventually actually allow flipping through pages, if there are more favorites than will fit on one page
+        let bottom_bar = BottomBar::new(rect![self.rect.min.x, self.rect.max.y - small_height + big_thickness,
+            self.rect.max.x, self.rect.max.y], 0, 1);
+        self.children.push(Box::new(bottom_bar) as Box<dyn View>);
+    }
+
+    fn create_marked_for_later(&mut self, top_pos: i32) {
+        let marked_for_later = Fave::new(
+            self.rect, top_pos,
+            "Marked For Later".to_string(),
+            Event::LoadHistory(super::works::HistoryView::MarkedForLater));
+
+        self.children.push(Box::new(marked_for_later) as Box<dyn View>);
+    }
+
+    fn create_fav_search(&mut self, fave: (String, Url), top_pos: i32) {
+        let fave = Fave::new(
+            self.rect, top_pos,
+            (*fave.0).to_string(),
+            Event::LoadIndex((fave.1).to_string()));
+
+        self.children.push(Box::new(fave) as Box<dyn View>);
+    }
+
+    fn set_shelf_index(&mut self, index: usize) {
+        self.shelf_index = index;
+    }
+
+    fn open_search_bar(&mut self, keyboard_layouts: &BTreeMap<String, Layout>, keyboard_name: String, rq: &mut RenderQueue) {
```

Comment:
Combined the toggle_search and toggle_keyboard logic into a single function, and then split it back out into open_search_bar and close_search_bar.  open_search_bar displays both the search bar and the keyboard, and close_search_bar removes them both